include/alarm_heap.hpp

```diff
@@ -0,0 +1,198 @@
+* under which the OpenSSL Project distributes the OpenSSL toolkit software,
+* as those licenses appear in the file LICENSE-OPENSSL.
+*/
+
+#ifndef ALARM_HEAP_HPP
+#define ALARM_HEAP_HPP
+
+#include <time.h>
+
+#include <string>
+#include <map>
+#include <cstdint>
+
+#include "alarm_table_defs.hpp"
+#include "alarm_trap_sender.hpp"
+#include "timer_heap.h"
+#include "utils.h"
+#include "cond_var.h"
+
+enum AlarmSeverityChange
+{
+  NO_CHANGE,
+  INCREASED_SEVERITY,
+  REDUCED_SEVERITY,
+};
+
+
+class AlarmInHeap : public HeapableTimer
+{
+public:
+  AlarmInHeap(unsigned int index) :
+    _pop_time(UINT64_MAX),
+    _severity(AlarmDef::Severity::UNDEFINED_SEVERITY),
+    _index(index)
+ {}
+
+  // Updates the pop time of the AlarmInHeap, and rebalances the heap to
+  // account for the changed pop time
+  void update_pop_time(uint64_t time_to_add)
+  {
+    _pop_time = Utils::get_time() + time_to_add;
+    _heap->rebalance(this);
+  }
+
+  // Removes the AlarmInHeap from the heap, and sets the pop time/severity
+  // to clearly invalid values (that won't cause INFORMS to be sent in error
+  // cases)
+  void remove_from_heap()
+  {
+    _pop_time = UINT64_MAX;
+    _severity = AlarmDef::Severity::UNDEFINED_SEVERITY;
+    _heap->remove(this);
+  }
+
+  const unsigned int index() { return _index; }
+  AlarmDef::Severity severity() { return _severity; }
+  void set_severity(AlarmDef::Severity severity) { _severity = severity; }
+  bool in_heap() { return (_heap != nullptr); }
+  uint64_t get_pop_time() const { return _pop_time; }
+
+private:
+  uint64_t _pop_time;
+  AlarmDef::Severity _severity;
+  unsigned int _index;
+};
+
+class AlarmInfo
+{
+public:
+  AlarmInfo(AlarmInHeap* alarm_in_heap,
+            AlarmDef::Severity severity) :
+    _alarm_in_heap(alarm_in_heap),
+    _severity(severity)
+  {}
+
+  ~AlarmInfo();
+
+  // Compares a new severity with the current severity for this alarm,
+  // and returns the severity change
+  // @params severity - the new severity
+  // @returns         - the severity change
+  AlarmSeverityChange new_alarm_severity(AlarmDef::Severity severity);
+
+  // Updates the alarm in the heap's severity, and its time to pop
+  void update_alarm_in_heap(AlarmDef::Severity new_severity,
+                            uint64_t time_to_delay);
+
+  AlarmInHeap* alarm_in_heap() { return _alarm_in_heap; }
+  AlarmDef::Severity get_severity() { return _severity; }
+  void set_severity(AlarmTableDef& alarm_table_def);
+
+  // Determines whether this alarm should be resent after failure
+  bool should_resend_alarm(AlarmDef::Severity severity)
+  {
+    // Resend the alarm if its severity and the current severity match,
+    // and there isn't already an alarm in the heap waiting to update
+    // the severity.
+    return ((_severity == severity) && (!_alarm_in_heap->in_heap()));
+  }
+
+private:
+  AlarmInHeap* _alarm_in_heap;
+  AlarmDef::Severity _severity;
+};
+
+typedef std::map<unsigned int, AlarmInfo*> AllAlarmsStateMap;
```

Comment:
I'm not sure this typedef makes things much clearer. (Pedantically, the fact that this contains *all* alarms isn't type system information.)

A typedef that might be good is `typedef unsigned int AlarmIndex`. `std::map<AlarmIndex, unsigned int>& dup_check` is clearer than `std::map<unsigned int, unsigned int>& dup_check`, for example.